### PR TITLE
Add calculation of total information content and information temperature

### LINKED ISSUE #5011: Add calculation of total information content and information temperature
Extend information properties calculation and visualisation with two more fundamental functions:
- Total information content (or mutual information)
- Information temperature

# Definitions

Total information content or mutual information could be considered analogous to enthalpy in thermodynamics. For mutual information:
I(X;Y)=∑x∈X∑y∈Yp(x,y)log(p(x,y)/p(x)p(y))

-    p(x,y): Joint probability of X and Y
-    p(x), p(y): Marginal probabilities of X and Y

Note, that mutual information is only possible to calculate *relatively* to something else, thus not always applicable. Thus, what could be done is to calculate mutual information compared to somethiing else, e.g. to the random uniform sequence or two data blocks against each other.

In information theory, information temperature (Tinfo) is sometimes defined as the derivative of entropy with respect to an "informational energy" or complexity measure:
Tinfo=∂H/∂Einfo

-    H: Entropy
-    Einfo: Informational energy or complexity

# Entropy and mutual information relationship

-    In thermodynamics, entropy and enthalpy are connected through temperature and Gibbs free energy, describing the balance between disorder and energy.
-    In information theory, entropy quantifies uncertainty, while mutual information (an enthalpy-like concept) quantifies shared information. The analogy highlights how both fields use these concepts to describe system behavior, though the mathematical relationships differ.

# Applications
- Code similarity
- Data dependency calculation (e.g. between fields of a structure or file headers)
- Detecting patterns in obfuscated data and code

See example (generated by an AI):
```python
import numpy as np
from collections import Counter

def calculate_entropy(sequence):
    """Calculate Shannon entropy of a byte sequence."""
    length = len(sequence)
    counts = Counter(sequence)
    probabilities = np.array([count / length for count in counts.values()])
    entropy = -np.sum(probabilities * np.log2(probabilities))
    return entropy

def calculate_mutual_information(seq1, seq2):
    """Calculate mutual information between two sequences."""
    joint_counts = Counter(zip(seq1, seq2))
    joint_probabilities = np.array([count / len(seq1) for count in joint_counts.values()])
    seq1_counts = Counter(seq1)
    seq2_counts = Counter(seq2)
    seq1_probabilities = np.array([count / len(seq1) for count in seq1_counts.values()])
    seq2_probabilities = np.array([count / len(seq2) for count in seq2_counts.values()])

    # Calculate joint entropy
    joint_entropy = -np.sum(joint_probabilities * np.log2(joint_probabilities))

    # Calculate individual entropies
    entropy_seq1 = -np.sum(seq1_probabilities * np.log2(seq1_probabilities))
    entropy_seq2 = -np.sum(seq2_probabilities * np.log2(seq2_probabilities))

    # Mutual information
    mutual_information = entropy_seq1 + entropy_seq2 - joint_entropy
    return mutual_information

def calculate_information_temperature(entropy, sequence_length):
    """Approximate information temperature as the derivative of entropy with respect to length."""
    # Assuming entropy scales logarithmically with sequence length
    temperature = entropy / np.log2(sequence_length)
    return temperature

# Generate a random sequence of 1024 bytes
np.random.seed(42)  # For reproducibility
sequence = np.random.randint(0, 256, 1024, dtype=np.uint8)

# Calculate entropy
entropy = calculate_entropy(sequence)
print(f"Entropy: {entropy:.4f} bits")

# Split sequence into two halves and calculate mutual information
half_length = len(sequence) // 2
seq1, seq2 = sequence[:half_length], sequence[half_length:]
mutual_information = calculate_mutual_information(seq1, seq2)
print(f"Mutual Information (Enthalpy Analogy): {mutual_information:.4f} bits")

# Calculate information temperature
information_temperature = calculate_information_temperature(entropy, len(sequence))
print(f"Information Temperature: {information_temperature:.4f}")
```
And an output example:
```
Entropy: 7.8007 bits
Mutual Information (Enthalpy Analogy): 6.1521 bits
Information Temperature: 0.7801
```

See these sources for more context:
- https://arxiv.org/abs/cs/0701016
- https://arxiv.org/abs/2307.12841
- https://en.wikipedia.org/wiki/Mutual_information
- https://crypto.ethz.ch/publications/files/RenMau02a.pdf
- https://eprint.iacr.org/2007/198.pdf

### PR DESCRIPTION
**Pull request by vaguue**
# SQUASH ME

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository.
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style).
- [x] I've documented every `RZ_API` function and struct this PR changes.
- [x] I've added tests that prove my changes are effective (required for changes to `RZ_API`).
- [] I've updated the [Rizin book](https://github.com/rizinorg/book) with the relevant information (if needed).

**Detailed description**

This release adds temperature and mutual entropy
...

**Test plan**

Run
```meson test --print-errorlogs -C build hash```

...

**Closing issues**

closes #5011 

...

### CODE REVIEW

File: librz/hash/algorithms/temperature/temperature.c [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+		entropy /= log2((double)RZ_MIN(ctx->size, 256));
+	}
+        double temperature = 0.0;
+	if (ctx->size > 1) {
+		temperature = entropy / log2((double)ctx->size);
+		if (fraction) {
+			temperature *= log2((double)ctx->size)
+			                / log2((double)RZ_MIN(ctx->size, 256));
+		}
+	}
```
Comment (wargio) [inline librz/hash/algorithms/temperature/temperature.c]: invert this and return early.

Comment (vaguue) [reply librz/hash/algorithms/temperature/temperature.c]: done, if I've got you right. Also can you please check the comments I left in the original issue https://github.com/rizinorg/rizin/issues/5011, I've noticed you've implemented a lot of the functions in the hashing module

**Review by wargio — CHANGES_REQUESTED** [review]
When size is 1, log2 returns 0. And you can't divide by zero.

**Comment by vaguue** [comment]
> When size is 1, log2 returns 0. And you can't divide by zero.

But there is a 
```
        if (ctx->size <= 1) {
                rz_write_be_double(digest, 0.0);
                return true;
        }
```

check, I've added a test where the string has a length of 1

**Comment by wargio** [comment]
> check, I've added a test where the string has a length of 1

i must have missed it. then great! thanks for adding the test. that one was important, since it's an edge case.

File: librz/hash/algorithms/temperature/temperature.c [hunk 2]
```diff
@@ -0,0 +1,52 @@
+// SPDX-FileCopyrightText: 2025 Seva <little_scamp@yahoo.comt>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include "temperature.h"
```
Comment (notxvilka) [inline librz/hash/algorithms/temperature/temperature.c:20]: Please add `\file` Doxygen description on what temperature is.

File: librz/hash/hash.c [hunk 3]
```diff
@@ -127,6 +127,27 @@ RZ_API ut32 rz_hash_xxhash(RZ_NONNULL const ut8 *input, size_t size) {
 	return XXH32(input, size, 0);
 }
 
+/**
+ * \brief      Calculates the temperature of the given input
```
Comment (notxvilka) [inline librz/hash/hash.c]: `information temperature` to avoid any misunderstanding.

**Review by notxvilka — CHANGES_REQUESTED** [review]
Please fix remaining tests: https://github.com/rizinorg/rizin/actions/runs/15266026217/job/42960766803?pr=5167#step:18:926

You could use `rz-test -i` or `rz-test -y` modes to automate.

File: librz/hash/hash.c (hunk 3 above)
Comment (vaguue) [reply librz/hash/hash.c]: done

File: librz/hash/algorithms/temperature/temperature.c (hunk 2 above)
Comment (vaguue) [reply librz/hash/algorithms/temperature/temperature.c:20]: done, hope I got it right

File: librz/hash/algorithms/temperature/temperature.c [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+ * \f$ T_{\text{info}} = \frac{H}{\log_2 N}\f$
+ * where **H** is the Shannon entropy of the byte block and **N**
+ * is the block length in bytes.
+ *
+ * Key points
+ * ----------
+ * * Range     : 0 ≤ T<sub>info</sub> ≤ 1 for blocks up to 256 bytes
+ *   (with `fraction=true` the value is explicitly normalised to [0, 1]
+ *
+ * @author Seva <little_scamp@yahoo.com>
```
Comment (notxvilka) [inline librz/hash/algorithms/temperature/temperature.c]: We use `\author` and `\file` syntax

File: librz/hash/algorithms/temperature/temperature.c [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+ * @copyright 2025
```
Comment (notxvilka) [inline librz/hash/algorithms/temperature/temperature.c]: Copyright and license are specified in SPDX at the file's top, no need to duplicate them here.

**Review by notxvilka — CHANGES_REQUESTED** [review]
Please fix tests also

**Comment by notxvilka** [comment]
Please also fix `clang-format` errors:
```
test/unit/test_hash.c:108:74: error: code should be clang-formatted [-Wclang-format-violations]
        { INDATA("abcdefgh"), .algo = "temperature", .expected = "1.00000000" },
                                                                                ^
```

**Comment by vaguue** [comment]
> Please also fix `clang-format` errors:
> 
> ```
> test/unit/test_hash.c:108:74: error: code should be clang-formatted [-Wclang-format-violations]
>         { INDATA("abcdefgh"), .algo = "temperature", .expected = "1.00000000" },
>                                                                                 ^
> ```

I fixed this one, but there are several linter errors even in the code I basically copied, so I'm not sure what to fix and whatnot

File: librz/hash/algorithms/temperature/temperature.c [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+ * where **H** is the Shannon entropy of the byte block and **N**
+ * is the block length in bytes.
+ *
+ * Key points
+ * ----------
+ * * Range : 0 ≤ T<sub>info</sub> ≤ 1 for blocks ≤ 256 bytes.
+ *   If `fraction == true`, the value is additionally normalised to [0, 1]
+ *   exactly like “fractional entropy”.
+ *
+ * \author Seva <little_scamp@yahoo.com>
```
Comment (notxvilka) [inline librz/hash/algorithms/temperature/temperature.c]: Duplicate of SPDX

**Review by notxvilka — CHANGES_REQUESTED** [review]

**Comment by notxvilka** [comment]
> > Please also fix `clang-format` errors:
> > ```
> > test/unit/test_hash.c:108:74: error: code should be clang-formatted [-Wclang-format-violations]
> >         { INDATA("abcdefgh"), .algo = "temperature", .expected = "1.00000000" },
> >                                                                                 ^
> > ```
> 
> I fixed this one, but there are several linter errors even in the code I basically copied, so I'm not sure what to fix and whatnot

Looks like you have a lot of places where there is an empty space after the comma/printable character:
```
librz/hash/algorithms/temperature/temperature.c:42:46: error: code should be clang-formatted [-Wclang-format-violations]
        rz_return_val_if_fail(ctx && digest, false);
                                                    ^
librz/hash/algorithms/temperature/temperature.c:44:30: error: code should be clang-formatted [-Wclang-format-violations]
        if (ctx->size <= 1) {
                             ^
librz/hash/algorithms/temperature/temperature.c:45:49: error: code should be clang-formatted [-Wclang-format-violations]
                rz_write_be_double(digest, 0.0);
                                                ^
librz/hash/algorithms/temperature/temperature.c:46:29: error: code should be clang-formatted [-Wclang-format-violations]
                return true;
                            ^
librz/hash/algorithms/temperature/temperature.c:61:3: error: code should be clang-formatted [-Wclang-format-violations]
        }
         ^
librz/hash/algorithms/temperature/temperature.c:63:64: error: code should be clang-formatted [-Wclang-format-violations]
        double temperature = entropy / log2((double)ctx->size);
                                                               ^
librz/hash/algorithms/temperature/temperature.c:64:24: error: code should be clang-formatted [-Wclang-format-violations]
        if (fraction) {
                       ^
librz/hash/algorithms/temperature/temperature.c:65:57: error: code should be clang-formatted [-Wclang-format-violations]
                temperature *= log2((double)ctx->size) /
                                                        ^
librz/hash/algorithms/temperature/temperature.c:66:69: error: code should be clang-formatted [-Wclang-format-violations]
                               log2((double)RZ_MIN(ctx->size, 256));
                                                                    ^
librz/hash/algorithms/temperature/temperature.c:67:10: error: code should be clang-formatted [-Wclang-format-violations]
        }
         ^
librz/hash/algorithms/temperature/temperature.c:68:49: error: code should be clang-formatted [-Wclang-format-violations]
        rz_write_be_double(digest, temperature);
                                                ^
```

Just setup your editor/IDE to show them or remove automatically.

File: librz/hash/algorithms/temperature/temperature.c (hunk 6 above)
Comment (vaguue) [reply librz/hash/algorithms/temperature/temperature.c]: Sorry for that, fixed it

**Comment by vaguue** [comment]
> Please fix tests also

done

**Review by notxvilka — APPROVED** [review]
